src/CanvasContainer.tsx

```diff
@@ -2,171 +2,400 @@ import React, { CSSProperties, useEffect, useRef } from 'react';
 import { canvasModel, ZoomFactor } from './canvasMachine';
 import { useCanvas } from './CanvasContext';
 import { useMachine } from '@xstate/react';
+import {
+  assign,
+  actions,
+  sendParent,
+  ContextFrom,
+  SpecialTargets,
+} from 'xstate';
 import { createModel } from 'xstate/lib/model';
 import { Point } from './pathUtils';
-import { AnyState, EmbedContext } from './types';
+import { isAcceptingSpaceNatively, isWithPlatformMetaKey } from './utils';
 import { useEmbed } from './embedContext';
-import { isWithPlatformMetaKey, isTextInputLikeElement } from './utils';
+import { AnyState } from './types';
 
-const dragModel = createModel(
+interface DragSession {
+  pointerId: number;
+  point: Point;
+}
+
+interface PointDelta {
+  x: number;
+  y: number;
+}
+
+const dragSessionModel = createModel(
   {
-    startPoint: { x: 0, y: 0 },
-    dragPoint: { x: 0, y: 0 },
-    dx: 0,
-    dy: 0,
-    embed: undefined as EmbedContext | undefined,
+    session: null as DragSession | null,
+    ref: null as React.MutableRefObject<HTMLElement> | null,
   },
   {
     events: {
-      LOCK: () => ({}),
-      RELEASE: () => ({}),
-      GRAB: (point: Point) => ({ point }),
-      DRAG: (point: Point) => ({ point }),
-      UNGRAB: () => ({}),
+      DRAG_SESSION_STARTED: ({ pointerId, point }: DragSession) => ({
+        pointerId,
+        point,
+      }),
+      DRAG_SESSION_STOPPED: () => ({}),
+      DRAG_POINT_MOVED: ({ point }: Pick<DragSession, 'point'>) => ({ point }),
     },
   },
 );
 
-const dragMachine = dragModel.createMachine({
-  initial: 'checking_embedded_mode',
-  states: {
-    checking_embedded_mode: {
-      always: [
-        {
-          target: 'in_embedded_mode',
-          cond: (ctx) => !!ctx.embed?.isEmbedded && !ctx.embed.pan,
+const dragSessionTracker = dragSessionModel.createMachine(
+  {
+    preserveActionOrder: true,
+    initial: 'idle',
+    states: {
+      idle: {
+        invoke: {
+          id: 'dragSessionStartedListener',
+          src:
+            ({ ref }) =>
+            (sendBack) => {
+              const node = ref!.current!;
+              const listener = (ev: PointerEvent) => {
+                const isLeftButton = ev.button === 0;
```

Comment:
```suggestion
                const isLeftMouseButton = ev.button === 0;
```

Reply:
is this check against `ev.buttton` consistent for other types of input other than a physical mouse?

Reply:
It is consistent in terms of the desired action - I think we should always assume that those are intentions to drag (for touch and pen this means "contact": https://w3c.github.io/pointerevents/#the-button-property ).

The variable might be named slightly inaccurate though - do you think I should rename it? How should I name it? `isLeftMouseButtonOrContact`? Or maybe I should just inline the condition to avoid naming stuff 😅

Reply:
Hard to find a name that satisfies all the inputs. Maybe it's fine to only name it for the mouse and move on. FWIW, `isLeftButton` is not readable to me.